packages/core/src/index.ts

```diff
@@ -313,10 +318,26 @@ export function buildHooksSystem<Properties = Record<string, unknown>>(
     }
 
     function css(
-      properties: WithHooks<HookProperties, Properties>,
+      ...properties: [
+        WithHooks<HookProperties, Properties>,
+        ...(WithHooks<HookProperties, Properties> | undefined)[],
+      ]
     ): Properties {
-      return cssImpl(
-        JSON.parse(JSON.stringify(properties)) as typeof properties,
+      return (
+        (JSON.parse(JSON.stringify(properties)) as typeof properties).filter(
+          x => x,
+        ) as WithHooks<HookProperties, Properties>[]
+      ).reduce(
+        (a, b) =>
+          cssImpl({
+            ...Object.fromEntries(
+              Object.entries(
+                cssImpl(a) as WithHooks<HookProperties, Properties>,
+              ).filter(([k]) => !(k in b)),
+            ),
+            ...b,
```

Comment:
I'm curious to know the perf impact of this change. [Spreading inside reduce is known to cause a performance hit](https://www.richsnapp.com/article/2019/06-09-reduce-spread-anti-pattern), and the nested `fromEntries`/`entries` calls seem like they could cause a noticeable slow down.

Might be good to test: https://perf.link/

Reply:
Hey, thanks @atomiks! I rewrote that bit in a more imperative style, and it does indeed appear to be faster! I'll push the update shortly.

https://perf.link/#eyJpZCI6ImF2bzQyYnRodmpxIiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImZ1bmN0aW9uIGlzSG9va1NwZWMoeCkge1xuICAgIGlmICgheCkge1xuICAgICAgICByZXR1cm4gZmFsc2U7XG4gICAgfVxuICAgIGlmICh0eXBlb2YgeCA9PT0gXCJzdHJpbmdcIikge1xuICAgICAgICByZXR1cm4gKHguc3RhcnRzV2l0aChcIjpcIikgfHxcbiAgICAgICAgICAgIHguc3RhcnRzV2l0aChcIkBtZWRpYSBcIikgfHxcbiAgICAgICAgICAgIHguc3RhcnRzV2l0aChcIkBjb250YWluZXIgXCIpIHx8XG4gICAgICAgICAgICB4LmluY2x1ZGVzKFwiJlwiKSk7XG4gICAgfVxuICAgIGlmICh0eXBlb2YgeCA9PT0gXCJvYmplY3RcIikge1xuICAgICAgICBpZiAoXCJvclwiIGluIHggJiYgeC5vciBpbnN0YW5jZW9mIEFycmF5KSB7XG4gICAgICAgICAgICByZXR1cm4gIXgub3Iuc29tZSh4eCA9PiAhaXNIb29rU3BlYyh4eCkpO1xuICAgICAgICB9XG4gICAgICAgIGlmIChcImFuZFwiIGluIHggJiYgeC5hbmQgaW5zdGFuY2VvZiBBcnJheSkge1xuICAgICAgICAgICAgcmV0dXJuICF4LmFuZC5zb21lKHh4ID0%2BICFpc0hvb2tTcGVjKHh4KSk7XG4gICAgICAgIH1cbiAgICB9XG4gICAgcmV0dXJuIGZhbHNlO1xufVxuLyoqIEBpbnRlcm5hbCAqL1xuZnVuY3Rpb24gZ2VuZXJpY1N0cmluZ2lmeShfLCB2YWx1ZSkge1xuICAgIGlmICh0eXBlb2YgdmFsdWUgPT09IFwic3RyaW5nXCIpIHtcbiAgICAgICAgcmV0dXJuIHZhbHVlO1xuICAgIH1cbiAgICBpZiAodHlwZW9mIHZhbHVlID09PSBcIm51bWJlclwiKSB7XG4gICAgICAgIHJldHVybiBgJHt2YWx1ZX1gO1xuICAgIH1cbiAgICByZXR1cm4gbnVsbDtcbn1cbmZ1bmN0aW9uIGhhc2gob2JqKSB7XG4gICAgY29uc3QganNvblN0cmluZyA9IEpTT04uc3RyaW5naWZ5KG9iaik7XG4gICAgbGV0IGhhc2hWYWx1ZSA9IDA7XG4gICAgZm9yIChsZXQgaSA9IDA7IGkgPCBqc29uU3RyaW5nLmxlbmd0aDsgaSsrKSB7XG4gICAgICAgIGNvbnN0IGNoYXJDb2RlID0ganNvblN0cmluZy5jaGFyQ29kZUF0KGkpO1xuICAgICAgICBoYXNoVmFsdWUgPSAoaGFzaFZhbHVlIDw8IDUpIC0gaGFzaFZhbHVlICsgY2hhckNvZGU7XG4gICAgICAgIGhhc2hWYWx1ZSAmPSAweDdmZmZmZmZmO1xuICAgIH1cbiAgICByZXR1cm4gaGFzaFZhbHVlLnRvU3RyaW5nKDM2KTtcbn1cblxuZnVuY3Rpb24gYnVpbGRIb29rc1N5c3RlbUEoc3RyaW5naWZ5ID0gZ2VuZXJpY1N0cmluZ2lmeSkge1xuICAgIHJldHVybiBmdW5jdGlvbiBjcmVhdGVIb29rcyhjb25maWcsIG9wdGlvbnMpIHtcbiAgICAgICAgY29uc3QgZmFsbGJhY2tLZXl3b3JkID0gKG9wdGlvbnMgPT09IG51bGwgfHwgb3B0aW9ucyA9PT0gdm9pZCAwID8gdm9pZCAwIDogb3B0aW9ucy5mYWxsYmFjaykgfHwgXCJ1bnNldFwiO1xuICAgICAgICBjb25zdCBzdHJpbmdpZnlJbXBsID0gKHByb3BlcnR5TmFtZSwgdmFsdWUpID0%2BIHtcbiAgICAgICAgICAgIHJldHVybiB0eXBlb2YgdmFsdWUgPT09IFwic3RyaW5nXCIgJiYgdmFsdWUuc3RhcnRzV2l0aChcInZhcihcIilcbiAgICAgICAgICAgICAgICA%2FIHZhbHVlXG4gICAgICAgICAgICAgICAgOiBzdHJpbmdpZnkocHJvcGVydHlOYW1lLCB2YWx1ZSk7XG4gICAgICAgIH07XG4gICAgICAgIGNvbnN0IGhvb2tJZCA9IChvcHRpb25zID09PSBudWxsIHx8IG9wdGlvbnMgPT09IHZvaWQgMCA%2FIHZvaWQgMCA6IG9wdGlvbnMuaG9va05hbWVUb0lkKSB8fFxuICAgICAgICAgICAgKChob29rTmFtZSkgPT4ge1xuICAgICAgICAgICAgICAgIGNvbnN0IHNwZWNIYXNoID0gaGFzaChjb25maWdbaG9va05hbWVdKTtcbiAgICAgICAgICAgICAgICByZXR1cm4gKG9wdGlvbnMgPT09IG51bGwgfHwgb3B0aW9ucyA9PT0gdm9pZCAwID8gdm9pZCAwIDogb3B0aW9ucy5kZWJ1ZylcbiAgICAgICAgICAgICAgICAgICAgPyBgJHtob29rTmFtZS5yZXBsYWNlKC9bXkEtWmEtejAtOS1dL2csIFwiX1wiKX0tJHtzcGVjSGFzaH1gXG4gICAgICAgICAgICAgICAgICAgIDogc3BlY0hhc2g7XG4gICAgICAgICAgICB9KTtcbiAgICAgICAgZnVuY3Rpb24gZm9yRWFjaEhvb2socHJvcGVydGllcywgY2FsbGJhY2spIHtcbiAgICAgICAgICAgIHJldHVybiBPYmplY3QuZW50cmllcyhwcm9wZXJ0aWVzKVxuICAgICAgICAgICAgICAgIC5maWx0ZXIoKFtrZXldKSA9PiBrZXkgaW4gY29uZmlnKVxuICAgICAgICAgICAgICAgIC5mb3JFYWNoKChbaG9va05hbWUsIHZhbHVlXSkgPT4ge1xuICAgICAgICAgICAgICAgIGNhbGxiYWNrKGhvb2tOYW1lLCB2YWx1ZSk7XG4gICAgICAgICAgICB9KTtcbiAgICAgICAgfVxuICAgICAgICBjb25zdCBob29rcyA9IE9iamVjdC5lbnRyaWVzKGNvbmZpZylcbiAgICAgICAgICAgIC5tYXAoKFtuYW1lLCBkZWZpbml0aW9uXSkgPT4ge1xuICAgICAgICAgICAgZnVuY3Rpb24gbmVzdChpbnB1dCkge1xuICAgICAgICAgICAgICAgIGlmICh0eXBlb2YgaW5wdXQgPT09IFwib2JqZWN0XCIpIHtcbiAgICAgICAgICAgICAgICAgICAgaWYgKFwiYW5kXCIgaW4gaW5wdXQpIHtcbiAgICAgICAgICAgICAgICAgICAgICAgIGlmIChpbnB1dC5hbmQubGVuZ3RoID4gMikge1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgIGNvbnN0IFtsZWZ0LCAuLi5yZXN0XSA9IGlucHV0LmFuZDtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICByZXR1cm4geyBhbmQ6IFtsZWZ0LCBuZXN0KHsgYW5kOiByZXN0IH0pXSB9O1xuICAgICAgICAgICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgICAgICAgICAgcmV0dXJuIHtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICBhbmQ6IGlucHV0LmFuZC5tYXAoaXRlbSA9PiB0eXBlb2YgaXRlbSA9PT0gXCJzdHJpbmdcIiA%2FIGl0ZW0gOiBuZXN0KGl0ZW0pKSxcbiAgICAgICAgICAgICAgICAgICAgICAgIH07XG4gICAgICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICAgICAgaWYgKFwib3JcIiBpbiBpbnB1dCkge1xuICAgICAgICAgICAgICAgICAgICAgICAgaWYgKGlucHV0Lm9yLmxlbmd0aCA%2BIDIpIHtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICBjb25zdCBbbGVmdCwgLi4ucmVzdF0gPSBpbnB1dC5vcjtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICByZXR1cm4geyBvcjogW2xlZnQsIG5lc3QoeyBvcjogcmVzdCB9KV0gfTtcbiAgICAgICAgICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICAgICAgICAgIHJldHVybiB7XG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgb3I6IGlucHV0Lm9yLm1hcChpdGVtID0%2BIHR5cGVvZiBpdGVtID09PSBcInN0cmluZ1wiID8gaXRlbSA6IG5lc3QoaXRlbSkpLFxuICAgICAgICAgICAgICAgICAgICAgICAgfTtcbiAgICAgICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICByZXR1cm4gaW5wdXQ7XG4gICAgICAgICAgICB9XG4gICAgICAgICAgICBpZiAoIWlzSG9va1NwZWMoZGVmaW5pdGlvbikgfHwgdHlwZW9mIGRlZmluaXRpb24gIT09IFwib2JqZWN0XCIpIHtcbiAgICAgICAgICAgICAgICByZXR1cm4gW25hbWUsIGRlZmluaXRpb25dO1xuICAgICAgICAgICAgfVxuICAgICAgICAgICAgcmV0dXJuIFtuYW1lLCBuZXN0KGRlZmluaXRpb24pXTtcbiAgICAgICAgfSlcbiAgICAgICAgICAgIC5mbGF0TWFwKChbbmFtZSwgZGVmaW5pdGlvbl0pID0%2BIChmdW5jdGlvbiBob29rc0NTUyhuYW1lLCBkZWZpbml0aW9uKSB7XG4gICAgICAgICAgICBpZiAoIWlzSG9va1NwZWMoZGVmaW5pdGlvbikpIHtcbiAgICAgICAgICAgICAgICByZXR1cm4gW107XG4gICAgICAgICAgICB9XG4gICAgICAgICAgICBpZiAodHlwZW9mIGRlZmluaXRpb24gPT09IFwib2JqZWN0XCIpIHtcbiAgICAgICAgICAgICAgICBsZXQgYSwgb3BlcmF0b3IsIGIsIGV4dHJhSG9va3NDU1MgPSBbXTtcbiAgICAgICAgICAgICAgICBpZiAoXCJvclwiIGluIGRlZmluaXRpb24pIHtcbiAgICAgICAgICAgICAgICAgICAgb3BlcmF0b3IgPSBcIm9yXCI7XG4gICAgICAgICAgICAgICAgICAgIFthLCBiXSA9IGRlZmluaXRpb24ub3I7XG4gICAgICAgICAgICAgICAgICAgIGlmICghYSkge1xuICAgICAgICAgICAgICAgICAgICAgICAgcmV0dXJuIFtdO1xuICAgICAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgICAgIGlmICghYikge1xuICAgICAgICAgICAgICAgICAgICAgICAgcmV0dXJuIGhvb2tzQ1NTKG5hbWUsIGEpO1xuICAgICAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgICAgIGV4dHJhSG9va3NDU1MgPSBbXG4gICAgICAgICAgICAgICAgICAgICAgICB7XG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgaW5pdDogKGZ1bmN0aW9uIGFvcmIoeCkge1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICBjb25zdCBhID0gYCR7eH1BYDtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgY29uc3QgYiA9IGAke3h9QmA7XG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIHJldHVybiBbXG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICBgLS0ke3h9LTA6dmFyKC0tJHthfS0wLHZhcigtLSR7Yn0tMCkpO2AsXG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICBgLS0ke3h9LTE6dmFyKC0tJHthfS0xKSB2YXIoLS0ke2J9LTEpO2AsXG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIF0uam9pbihcIlwiKTtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICB9KShuYW1lKSxcbiAgICAgICAgICAgICAgICAgICAgICAgIH0sXG4gICAgICAgICAgICAgICAgICAgIF07XG4gICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgIGVsc2UgaWYgKFwiYW5kXCIgaW4gZGVmaW5pdGlvbikge1xuICAgICAgICAgICAgICAgICAgICBvcGVyYXRvciA9IFwiYW5kXCI7XG4gICAgICAgICAgICAgICAgICAgIFthLCBiXSA9IGRlZmluaXRpb24uYW5kO1xuICAgICAgICAgICAgICAgICAgICBpZiAoIWEpIHtcbiAgICAgICAgICAgICAgICAgICAgICAgIHJldHVybiBbXTtcbiAgICAgICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgICAgICBpZiAoIWIpIHtcbiAgICAgICAgICAgICAgICAgICAgICAgIHJldHVybiBob29rc0NTUyhuYW1lLCBhKTtcbiAgICAgICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgICAgICBleHRyYUhvb2tzQ1NTID0gW1xuICAgICAgICAgICAgICAgICAgICAgICAge1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgIGluaXQ6IChmdW5jdGlvbiBhYW5kYih4KSB7XG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIGNvbnN0IGEgPSBgJHt4fUFgO1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICBjb25zdCBiID0gYCR7eH1CYDtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgcmV0dXJuIFtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIGAtLSR7eH0tMDp2YXIoLS0ke2F9LTApIHZhcigtLSR7Yn0tMCk7YCxcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIGAtLSR7eH0tMTp2YXIoLS0ke2F9LTEsdmFyKC0tJHtifS0xKSk7YCxcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgXS5qb2luKFwiXCIpO1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgIH0pKG5hbWUpLFxuICAgICAgICAgICAgICAgICAgICAgICAgfSxcbiAgICAgICAgICAgICAgICAgICAgXTtcbiAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgaWYgKG9wZXJhdG9yKSB7XG4gICAgICAgICAgICAgICAgICAgIHJldHVybiBbXG4gICAgICAgICAgICAgICAgICAgICAgICAuLi5ob29rc0NTUyhgJHtuYW1lfUFgLCBhKSxcbiAgICAgICAgICAgICAgICAgICAgICAgIC4uLmhvb2tzQ1NTKGAke25hbWV9QmAsIGIpLFxuICAgICAgICAgICAgICAgICAgICAgICAgLi4uZXh0cmFIb29rc0NTUyxcbiAgICAgICAgICAgICAgICAgICAgXTtcbiAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICB9XG4gICAgICAgICAgICBjb25zdCBpbml0ID0gYC0tJHtuYW1lfS0wOmluaXRpYWw7LS0ke25hbWV9LTE6IDtgO1xuICAgICAgICAgICAgbGV0IHJ1bGU7XG4gICAgICAgICAgICBpZiAodHlwZW9mIGRlZmluaXRpb24gPT09IFwic3RyaW5nXCIpIHtcbiAgICAgICAgICAgICAgICBpZiAoZGVmaW5pdGlvbi5pbmNsdWRlcyhcIiZcIikpIHtcbiAgICAgICAgICAgICAgICAgICAgcnVsZSA9IGAke2RlZmluaXRpb24ucmVwbGFjZSgvJi9nLCBcIipcIil9ey0tJHtuYW1lfS0wOiA7LS0ke25hbWV9LTE6aW5pdGlhbDt9YDtcbiAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgZWxzZSBpZiAoZGVmaW5pdGlvbi5zdGFydHNXaXRoKFwiOlwiKSkge1xuICAgICAgICAgICAgICAgICAgICBydWxlID0gYCR7ZGVmaW5pdGlvbn17LS0ke25hbWV9LTA6IDstLSR7bmFtZX0tMTppbml0aWFsO31gO1xuICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICBlbHNlIGlmIChkZWZpbml0aW9uLnN0YXJ0c1dpdGgoXCJAXCIpKSB7XG4gICAgICAgICAgICAgICAgICAgIHJ1bGUgPSBgJHtkZWZpbml0aW9ufXsqey0tJHtuYW1lfS0wOiA7LS0ke25hbWV9LTE6aW5pdGlhbDt9fWA7XG4gICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgfVxuICAgICAgICAgICAgcmV0dXJuIHJ1bGUgPT09IHVuZGVmaW5lZCA%2FIFtdIDogW3sgaW5pdCwgcnVsZSB9XTtcbiAgICAgICAgfSkoaG9va0lkKG5hbWUpLCBkZWZpbml0aW9uKSlcbiAgICAgICAgICAgIC5yZWR1Y2UoKGFjYywgeyBpbml0ID0gXCJcIiwgcnVsZSA9IFwiXCIgfSkgPT4gKHtcbiAgICAgICAgICAgIGluaXQ6IGFjYy5pbml0ICsgaW5pdCxcbiAgICAgICAgICAgIHJ1bGU6IGFjYy5ydWxlICsgcnVsZSxcbiAgICAgICAgfSksIHtcbiAgICAgICAgICAgIGluaXQ6IFwiXCIsXG4gICAgICAgICAgICBydWxlOiBcIlwiLFxuICAgICAgICB9KTtcbiAgICAgICAgZnVuY3Rpb24gY3NzSW1wbChwcm9wZXJ0aWVzLCBmYWxsYmFjayA9IHByb3BlcnR5TmFtZSA9PiBzdHJpbmdpZnlJbXBsKHByb3BlcnR5TmFtZSwgcHJvcGVydGllc1twcm9wZXJ0eU5hbWVdKSkge1xuICAgICAgICAgICAgY29uc3Qga2V5cyA9IE9iamVjdC5rZXlzKHByb3BlcnRpZXMpO1xuICAgICAgICAgICAgZm9yRWFjaEhvb2socHJvcGVydGllcywgKGhvb2tOYW1lLCBpbm5lclByb3BlcnRpZXMpID0%2BIHtcbiAgICAgICAgICAgICAgICBjc3NJbXBsKGlubmVyUHJvcGVydGllcywgcHJvcGVydHlOYW1lID0%2BIHtcbiAgICAgICAgICAgICAgICAgICAgbGV0IHYgPSBzdHJpbmdpZnlJbXBsKHByb3BlcnR5TmFtZSwgaW5uZXJQcm9wZXJ0aWVzW3Byb3BlcnR5TmFtZV0pO1xuICAgICAgICAgICAgICAgICAgICBpZiAodiA9PT0gbnVsbCkge1xuICAgICAgICAgICAgICAgICAgICAgICAgdiA9IGZhbGxiYWNrKHByb3BlcnR5TmFtZSk7XG4gICAgICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICAgICAgaWYgKHYgPT09IG51bGwpIHtcbiAgICAgICAgICAgICAgICAgICAgICAgIHYgPSBmYWxsYmFja0tleXdvcmQ7XG4gICAgICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICAgICAgcmV0dXJuIHY7XG4gICAgICAgICAgICAgICAgfSk7XG4gICAgICAgICAgICAgICAgZm9yIChjb25zdCBwcm9wZXJ0eU5hbWVTdHIgaW4gaW5uZXJQcm9wZXJ0aWVzKSB7XG4gICAgICAgICAgICAgICAgICAgIGlmIChrZXlzLmluZGV4T2YocHJvcGVydHlOYW1lU3RyKSA%2BIGtleXMuaW5kZXhPZihob29rTmFtZSkpIHtcbiAgICAgICAgICAgICAgICAgICAgICAgIGNvbnRpbnVlO1xuICAgICAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgICAgIGNvbnN0IHByb3BlcnR5TmFtZSA9IHByb3BlcnR5TmFtZVN0cjtcbiAgICAgICAgICAgICAgICAgICAgY29uc3QgdjEgPSBzdHJpbmdpZnlJbXBsKHByb3BlcnR5TmFtZSwgaW5uZXJQcm9wZXJ0aWVzW3Byb3BlcnR5TmFtZV0pO1xuICAgICAgICAgICAgICAgICAgICBpZiAodjEgIT09IG51bGwpIHtcbiAgICAgICAgICAgICAgICAgICAgICAgIGxldCB2MCA9IGZhbGxiYWNrKHByb3BlcnR5TmFtZSk7XG4gICAgICAgICAgICAgICAgICAgICAgICBpZiAodjAgPT09IG51bGwpIHtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICB2MCA9IGZhbGxiYWNrS2V5d29yZDtcbiAgICAgICAgICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICAgICAgICAgIGRlbGV0ZSBwcm9wZXJ0aWVzW3Byb3BlcnR5TmFtZV07XG4gICAgICAgICAgICAgICAgICAgICAgICBwcm9wZXJ0aWVzW3Byb3BlcnR5TmFtZV0gPSBgdmFyKC0tJHtob29rSWQoaG9va05hbWUpfS0xLCAke3YxfSkgdmFyKC0tJHtob29rSWQoaG9va05hbWUpfS0wLCAke3YwfSlgO1xuICAgICAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgIGRlbGV0ZSBwcm9wZXJ0aWVzW2hvb2tOYW1lXTtcbiAgICAgICAgICAgIH0pO1xuICAgICAgICAgICAgcmV0dXJuIHByb3BlcnRpZXM7XG4gICAgICAgIH1cbiAgICAgICAgZnVuY3Rpb24gY3NzKC4uLnByb3BlcnRpZXMpIHtcbiAgICAgICAgICAgIHJldHVybiBKU09OLnBhcnNlKEpTT04uc3RyaW5naWZ5KHByb3BlcnRpZXMpKS5maWx0ZXIoeCA9PiB4KS5yZWR1Y2UoKGEsIGIpID0%2BIGNzc0ltcGwoT2JqZWN0LmFzc2lnbihPYmplY3QuYXNzaWduKHt9LCBPYmplY3QuZnJvbUVudHJpZXMoT2JqZWN0LmVudHJpZXMoY3NzSW1wbChhKSkuZmlsdGVyKChba10pID0%2BICEoayBpbiBiKSkpKSwgYikpLCB7fSk7XG4gICAgICAgIH1cbiAgICAgICAgcmV0dXJuIFtgKnske2hvb2tzLmluaXR9fSR7aG9va3MucnVsZX1gLCBjc3NdO1xuICAgIH07XG59XG5cbmNvbnN0IGNyZWF0ZUhvb2tzQSA9IGJ1aWxkSG9va3NTeXN0ZW1BKCk7XG5jb25zdCBjc3NBID0gY3JlYXRlSG9va3NBKHtcbiAgYTogXCImLmFcIixcbiAgYjogXCImLmJcIixcbiAgYzogXCImLmNcIixcbiAgZDogXCImLmRcIixcbiAgZTogXCImLmVcIixcbiAgZjogXCImLmZcIixcbiAgZzogXCImLmdcIixcbiAgaDogXCImLmhcIixcbiAgaTogXCImLmlcIixcbiAgajogXCImLmpcIixcbiAgazogXCImLmtcIixcbiAgbDogXCImLmxcIixcbiAgbTogXCImLm1cIixcbiAgbjogXCImLm5cIixcbiAgbzogXCImLm9cIixcbiAgcDogXCImLnBcIixcbiAgcTogXCImLnFcIixcbiAgcjogXCImLnJcIixcbiAgczogXCImLnNcIixcbiAgdDogXCImLnRcIixcbiAgdTogXCImLnVcIixcbiAgdjogXCImLnZcIixcbiAgdzogXCImLndcIixcbiAgeDogXCImLnhcIixcbiAgeTogXCImLnlcIixcbiAgejogXCImLnpcIixcbn0pWzFdO1xuXG5mdW5jdGlvbiBidWlsZEhvb2tzU3lzdGVtQihzdHJpbmdpZnkgPSBnZW5lcmljU3RyaW5naWZ5KSB7XG4gICAgcmV0dXJuIGZ1bmN0aW9uIGNyZWF0ZUhvb2tzKGNvbmZpZywgb3B0aW9ucykge1xuICAgICAgICBjb25zdCBmYWxsYmFja0tleXdvcmQgPSAob3B0aW9ucyA9PT0gbnVsbCB8fCBvcHRpb25zID09PSB2b2lkIDAgPyB2b2lkIDAgOiBvcHRpb25zLmZhbGxiYWNrKSB8fCBcInVuc2V0XCI7XG4gICAgICAgIGNvbnN0IHN0cmluZ2lmeUltcGwgPSAocHJvcGVydHlOYW1lLCB2YWx1ZSkgPT4ge1xuICAgICAgICAgICAgcmV0dXJuIHR5cGVvZiB2YWx1ZSA9PT0gXCJzdHJpbmdcIiAmJiB2YWx1ZS5zdGFydHNXaXRoKFwidmFyKFwiKVxuICAgICAgICAgICAgICAgID8gdmFsdWVcbiAgICAgICAgICAgICAgICA6IHN0cmluZ2lmeShwcm9wZXJ0eU5hbWUsIHZhbHVlKTtcbiAgICAgICAgfTtcbiAgICAgICAgY29uc3QgaG9va0lkID0gKG9wdGlvbnMgPT09IG51bGwgfHwgb3B0aW9ucyA9PT0gdm9pZCAwID8gdm9pZCAwIDogb3B0aW9ucy5ob29rTmFtZVRvSWQpIHx8XG4gICAgICAgICAgICAoKGhvb2tOYW1lKSA9PiB7XG4gICAgICAgICAgICAgICAgY29uc3Qgc3BlY0hhc2ggPSBoYXNoKGNvbmZpZ1tob29rTmFtZV0pO1xuICAgICAgICAgICAgICAgIHJldHVybiAob3B0aW9ucyA9PT0gbnVsbCB8fCBvcHRpb25zID09PSB2b2lkIDAgPyB2b2lkIDAgOiBvcHRpb25zLmRlYnVnKVxuICAgICAgICAgICAgICAgICAgICA%2FIGAke2hvb2tOYW1lLnJlcGxhY2UoL1teQS1aYS16MC05LV0vZywgXCJfXCIpfS0ke3NwZWNIYXNofWBcbiAgICAgICAgICAgICAgICAgICAgOiBzcGVjSGFzaDtcbiAgICAgICAgICAgIH0pO1xuICAgICAgICBmdW5jdGlvbiBmb3JFYWNoSG9vayhwcm9wZXJ0aWVzLCBjYWxsYmFjaykge1xuICAgICAgICAgICAgcmV0dXJuIE9iamVjdC5lbnRyaWVzKHByb3BlcnRpZXMpXG4gICAgICAgICAgICAgICAgLmZpbHRlcigoW2tleV0pID0%2BIGtleSBpbiBjb25maWcpXG4gICAgICAgICAgICAgICAgLmZvckVhY2goKFtob29rTmFtZSwgdmFsdWVdKSA9PiB7XG4gICAgICAgICAgICAgICAgY2FsbGJhY2soaG9va05hbWUsIHZhbHVlKTtcbiAgICAgICAgICAgIH0pO1xuICAgICAgICB9XG4gICAgICAgIGNvbnN0IGhvb2tzID0gT2JqZWN0LmVudHJpZXMoY29uZmlnKVxuICAgICAgICAgICAgLm1hcCgoW25hbWUsIGRlZmluaXRpb25dKSA9PiB7XG4gICAgICAgICAgICBmdW5jdGlvbiBuZXN0KGlucHV0KSB7XG4gICAgICAgICAgICAgICAgaWYgKHR5cGVvZiBpbnB1dCA9PT0gXCJvYmplY3RcIikge1xuICAgICAgICAgICAgICAgICAgICBpZiAoXCJhbmRcIiBpbiBpbnB1dCkge1xuICAgICAgICAgICAgICAgICAgICAgICAgaWYgKGlucHV0LmFuZC5sZW5ndGggPiAyKSB7XG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgY29uc3QgW2xlZnQsIC4uLnJlc3RdID0gaW5wdXQuYW5kO1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgIHJldHVybiB7IGFuZDogW2xlZnQsIG5lc3QoeyBhbmQ6IHJlc3QgfSldIH07XG4gICAgICAgICAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgICAgICAgICByZXR1cm4ge1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgIGFuZDogaW5wdXQuYW5kLm1hcChpdGVtID0%2BIHR5cGVvZiBpdGVtID09PSBcInN0cmluZ1wiID8gaXRlbSA6IG5lc3QoaXRlbSkpLFxuICAgICAgICAgICAgICAgICAgICAgICAgfTtcbiAgICAgICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgICAgICBpZiAoXCJvclwiIGluIGlucHV0KSB7XG4gICAgICAgICAgICAgICAgICAgICAgICBpZiAoaW5wdXQub3IubGVuZ3RoID4gMikge1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgIGNvbnN0IFtsZWZ0LCAuLi5yZXN0XSA9IGlucHV0Lm9yO1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgIHJldHVybiB7IG9yOiBbbGVmdCwgbmVzdCh7IG9yOiByZXN0IH0pXSB9O1xuICAgICAgICAgICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgICAgICAgICAgcmV0dXJuIHtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICBvcjogaW5wdXQub3IubWFwKGl0ZW0gPT4gdHlwZW9mIGl0ZW0gPT09IFwic3RyaW5nXCIgPyBpdGVtIDogbmVzdChpdGVtKSksXG4gICAgICAgICAgICAgICAgICAgICAgICB9O1xuICAgICAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgIHJldHVybiBpbnB1dDtcbiAgICAgICAgICAgIH1cbiAgICAgICAgICAgIGlmICghaXNIb29rU3BlYyhkZWZpbml0aW9uKSB8fCB0eXBlb2YgZGVmaW5pdGlvbiAhPT0gXCJvYmplY3RcIikge1xuICAgICAgICAgICAgICAgIHJldHVybiBbbmFtZSwgZGVmaW5pdGlvbl07XG4gICAgICAgICAgICB9XG4gICAgICAgICAgICByZXR1cm4gW25hbWUsIG5lc3QoZGVmaW5pdGlvbildO1xuICAgICAgICB9KVxuICAgICAgICAgICAgLmZsYXRNYXAoKFtuYW1lLCBkZWZpbml0aW9uXSkgPT4gKGZ1bmN0aW9uIGhvb2tzQ1NTKG5hbWUsIGRlZmluaXRpb24pIHtcbiAgICAgICAgICAgIGlmICghaXNIb29rU3BlYyhkZWZpbml0aW9uKSkge1xuICAgICAgICAgICAgICAgIHJldHVybiBbXTtcbiAgICAgICAgICAgIH1cbiAgICAgICAgICAgIGlmICh0eXBlb2YgZGVmaW5pdGlvbiA9PT0gXCJvYmplY3RcIikge1xuICAgICAgICAgICAgICAgIGxldCBhLCBvcGVyYXRvciwgYiwgZXh0cmFIb29rc0NTUyA9IFtdO1xuICAgICAgICAgICAgICAgIGlmIChcIm9yXCIgaW4gZGVmaW5pdGlvbikge1xuICAgICAgICAgICAgICAgICAgICBvcGVyYXRvciA9IFwib3JcIjtcbiAgICAgICAgICAgICAgICAgICAgW2EsIGJdID0gZGVmaW5pdGlvbi5vcjtcbiAgICAgICAgICAgICAgICAgICAgaWYgKCFhKSB7XG4gICAgICAgICAgICAgICAgICAgICAgICByZXR1cm4gW107XG4gICAgICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICAgICAgaWYgKCFiKSB7XG4gICAgICAgICAgICAgICAgICAgICAgICByZXR1cm4gaG9va3NDU1MobmFtZSwgYSk7XG4gICAgICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICAgICAgZXh0cmFIb29rc0NTUyA9IFtcbiAgICAgICAgICAgICAgICAgICAgICAgIHtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICBpbml0OiAoZnVuY3Rpb24gYW9yYih4KSB7XG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIGNvbnN0IGEgPSBgJHt4fUFgO1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICBjb25zdCBiID0gYCR7eH1CYDtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgcmV0dXJuIFtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIGAtLSR7eH0tMDp2YXIoLS0ke2F9LTAsdmFyKC0tJHtifS0wKSk7YCxcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIGAtLSR7eH0tMTp2YXIoLS0ke2F9LTEpIHZhcigtLSR7Yn0tMSk7YCxcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgXS5qb2luKFwiXCIpO1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgIH0pKG5hbWUpLFxuICAgICAgICAgICAgICAgICAgICAgICAgfSxcbiAgICAgICAgICAgICAgICAgICAgXTtcbiAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgZWxzZSBpZiAoXCJhbmRcIiBpbiBkZWZpbml0aW9uKSB7XG4gICAgICAgICAgICAgICAgICAgIG9wZXJhdG9yID0gXCJhbmRcIjtcbiAgICAgICAgICAgICAgICAgICAgW2EsIGJdID0gZGVmaW5pdGlvbi5hbmQ7XG4gICAgICAgICAgICAgICAgICAgIGlmICghYSkge1xuICAgICAgICAgICAgICAgICAgICAgICAgcmV0dXJuIFtdO1xuICAgICAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgICAgIGlmICghYikge1xuICAgICAgICAgICAgICAgICAgICAgICAgcmV0dXJuIGhvb2tzQ1NTKG5hbWUsIGEpO1xuICAgICAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgICAgIGV4dHJhSG9va3NDU1MgPSBbXG4gICAgICAgICAgICAgICAgICAgICAgICB7XG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgaW5pdDogKGZ1bmN0aW9uIGFhbmRiKHgpIHtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgY29uc3QgYSA9IGAke3h9QWA7XG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIGNvbnN0IGIgPSBgJHt4fUJgO1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICByZXR1cm4gW1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgYC0tJHt4fS0wOnZhcigtLSR7YX0tMCkgdmFyKC0tJHtifS0wKTtgLFxuICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgYC0tJHt4fS0xOnZhcigtLSR7YX0tMSx2YXIoLS0ke2J9LTEpKTtgLFxuICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICBdLmpvaW4oXCJcIik7XG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgfSkobmFtZSksXG4gICAgICAgICAgICAgICAgICAgICAgICB9LFxuICAgICAgICAgICAgICAgICAgICBdO1xuICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICBpZiAob3BlcmF0b3IpIHtcbiAgICAgICAgICAgICAgICAgICAgcmV0dXJuIFtcbiAgICAgICAgICAgICAgICAgICAgICAgIC4uLmhvb2tzQ1NTKGAke25hbWV9QWAsIGEpLFxuICAgICAgICAgICAgICAgICAgICAgICAgLi4uaG9va3NDU1MoYCR7bmFtZX1CYCwgYiksXG4gICAgICAgICAgICAgICAgICAgICAgICAuLi5leHRyYUhvb2tzQ1NTLFxuICAgICAgICAgICAgICAgICAgICBdO1xuICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgIH1cbiAgICAgICAgICAgIGNvbnN0IGluaXQgPSBgLS0ke25hbWV9LTA6aW5pdGlhbDstLSR7bmFtZX0tMTogO2A7XG4gICAgICAgICAgICBsZXQgcnVsZTtcbiAgICAgICAgICAgIGlmICh0eXBlb2YgZGVmaW5pdGlvbiA9PT0gXCJzdHJpbmdcIikge1xuICAgICAgICAgICAgICAgIGlmIChkZWZpbml0aW9uLmluY2x1ZGVzKFwiJlwiKSkge1xuICAgICAgICAgICAgICAgICAgICBydWxlID0gYCR7ZGVmaW5pdGlvbi5yZXBsYWNlKC8mL2csIFwiKlwiKX17LS0ke25hbWV9LTA6IDstLSR7bmFtZX0tMTppbml0aWFsO31gO1xuICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICBlbHNlIGlmIChkZWZpbml0aW9uLnN0YXJ0c1dpdGgoXCI6XCIpKSB7XG4gICAgICAgICAgICAgICAgICAgIHJ1bGUgPSBgJHtkZWZpbml0aW9ufXstLSR7bmFtZX0tMDogOy0tJHtuYW1lfS0xOmluaXRpYWw7fWA7XG4gICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgIGVsc2UgaWYgKGRlZmluaXRpb24uc3RhcnRzV2l0aChcIkBcIikpIHtcbiAgICAgICAgICAgICAgICAgICAgcnVsZSA9IGAke2RlZmluaXRpb259eyp7LS0ke25hbWV9LTA6IDstLSR7bmFtZX0tMTppbml0aWFsO319YDtcbiAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICB9XG4gICAgICAgICAgICByZXR1cm4gcnVsZSA9PT0gdW5kZWZpbmVkID8gW10gOiBbeyBpbml0LCBydWxlIH1dO1xuICAgICAgICB9KShob29rSWQobmFtZSksIGRlZmluaXRpb24pKVxuICAgICAgICAgICAgLnJlZHVjZSgoYWNjLCB7IGluaXQgPSBcIlwiLCBydWxlID0gXCJcIiB9KSA9PiAoe1xuICAgICAgICAgICAgaW5pdDogYWNjLmluaXQgKyBpbml0LFxuICAgICAgICAgICAgcnVsZTogYWNjLnJ1bGUgKyBydWxlLFxuICAgICAgICB9KSwge1xuICAgICAgICAgICAgaW5pdDogXCJcIixcbiAgICAgICAgICAgIHJ1bGU6IFwiXCIsXG4gICAgICAgIH0pO1xuICAgICAgICBmdW5jdGlvbiBjc3NJbXBsKHByb3BlcnRpZXMsIGZhbGxiYWNrID0gcHJvcGVydHlOYW1lID0%2BIHN0cmluZ2lmeUltcGwocHJvcGVydHlOYW1lLCBwcm9wZXJ0aWVzW3Byb3BlcnR5TmFtZV0pKSB7XG4gICAgICAgICAgICBjb25zdCBrZXlzID0gT2JqZWN0LmtleXMocHJvcGVydGllcyk7XG4gICAgICAgICAgICBmb3JFYWNoSG9vayhwcm9wZXJ0aWVzLCAoaG9va05hbWUsIGlubmVyUHJvcGVydGllcykgPT4ge1xuICAgICAgICAgICAgICAgIGNzc0ltcGwoaW5uZXJQcm9wZXJ0aWVzLCBwcm9wZXJ0eU5hbWUgPT4ge1xuICAgICAgICAgICAgICAgICAgICBsZXQgdiA9IHN0cmluZ2lmeUltcGwocHJvcGVydHlOYW1lLCBpbm5lclByb3BlcnRpZXNbcHJvcGVydHlOYW1lXSk7XG4gICAgICAgICAgICAgICAgICAgIGlmICh2ID09PSBudWxsKSB7XG4gICAgICAgICAgICAgICAgICAgICAgICB2ID0gZmFsbGJhY2socHJvcGVydHlOYW1lKTtcbiAgICAgICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgICAgICBpZiAodiA9PT0gbnVsbCkge1xuICAgICAgICAgICAgICAgICAgICAgICAgdiA9IGZhbGxiYWNrS2V5d29yZDtcbiAgICAgICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgICAgICByZXR1cm4gdjtcbiAgICAgICAgICAgICAgICB9KTtcbiAgICAgICAgICAgICAgICBmb3IgKGNvbnN0IHByb3BlcnR5TmFtZVN0ciBpbiBpbm5lclByb3BlcnRpZXMpIHtcbiAgICAgICAgICAgICAgICAgICAgaWYgKGtleXMuaW5kZXhPZihwcm9wZXJ0eU5hbWVTdHIpID4ga2V5cy5pbmRleE9mKGhvb2tOYW1lKSkge1xuICAgICAgICAgICAgICAgICAgICAgICAgY29udGludWU7XG4gICAgICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICAgICAgY29uc3QgcHJvcGVydHlOYW1lID0gcHJvcGVydHlOYW1lU3RyO1xuICAgICAgICAgICAgICAgICAgICBjb25zdCB2MSA9IHN0cmluZ2lmeUltcGwocHJvcGVydHlOYW1lLCBpbm5lclByb3BlcnRpZXNbcHJvcGVydHlOYW1lXSk7XG4gICAgICAgICAgICAgICAgICAgIGlmICh2MSAhPT0gbnVsbCkge1xuICAgICAgICAgICAgICAgICAgICAgICAgbGV0IHYwID0gZmFsbGJhY2socHJvcGVydHlOYW1lKTtcbiAgICAgICAgICAgICAgICAgICAgICAgIGlmICh2MCA9PT0gbnVsbCkge1xuICAgICAgICAgICAgICAgICAgICAgICAgICAgIHYwID0gZmFsbGJhY2tLZXl3b3JkO1xuICAgICAgICAgICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgICAgICAgICAgZGVsZXRlIHByb3BlcnRpZXNbcHJvcGVydHlOYW1lXTtcbiAgICAgICAgICAgICAgICAgICAgICAgIHByb3BlcnRpZXNbcHJvcGVydHlOYW1lXSA9IGB2YXIoLS0ke2hvb2tJZChob29rTmFtZSl9LTEsICR7djF9KSB2YXIoLS0ke2hvb2tJZChob29rTmFtZSl9LTAsICR7djB9KWA7XG4gICAgICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICB9XG4gICAgICAgICAgICAgICAgZGVsZXRlIHByb3BlcnRpZXNbaG9va05hbWVdO1xuICAgICAgICAgICAgfSk7XG4gICAgICAgICAgICByZXR1cm4gcHJvcGVydGllcztcbiAgICAgICAgfVxuICAgICAgICBmdW5jdGlvbiBjc3MoLi4ucHJvcGVydGllcykge1xuICAgICAgICAgICAgY29uc3Qgc3R5bGVzID0gW3t9XS5jb25jYXQoSlNPTi5wYXJzZShKU09OLnN0cmluZ2lmeShwcm9wZXJ0aWVzKSkpO1xuICAgICAgICAgICAgZG8ge1xuICAgICAgICAgICAgICAgIGNvbnN0IGN1cnJlbnQgPSBzdHlsZXNbMF07XG4gICAgICAgICAgICAgICAgY29uc3QgbmV4dCA9IHN0eWxlc1sxXSB8fCB7fTtcbiAgICAgICAgICAgICAgICBmb3IgKGNvbnN0IGsgaW4gbmV4dCkge1xuICAgICAgICAgICAgICAgICAgICBpZiAoayBpbiBjdXJyZW50KSB7XG4gICAgICAgICAgICAgICAgICAgICAgICBkZWxldGUgY3VycmVudFtrXTtcbiAgICAgICAgICAgICAgICAgICAgfVxuICAgICAgICAgICAgICAgIH1cbiAgICAgICAgICAgICAgICBzdHlsZXNbMF0gPSBjc3NJbXBsKE9iamVjdC5hc3NpZ24oY3VycmVudCwgbmV4dCkpO1xuICAgICAgICAgICAgICAgIHN0eWxlcy5zcGxpY2UoMSwgMSk7XG4gICAgICAgICAgICB9IHdoaWxlIChzdHlsZXNbMV0pO1xuICAgICAgICAgICAgcmV0dXJuIHN0eWxlc1swXTtcbiAgICAgICAgfVxuICAgICAgICByZXR1cm4gW2AqeyR7aG9va3MuaW5pdH19JHtob29rcy5ydWxlfWAsIGNzc107XG4gICAgfTtcbn1cblxuY29uc3QgY3JlYXRlSG9va3NCID0gYnVpbGRIb29rc1N5c3RlbUIoKTtcbmNvbnN0IGNzc0IgPSBjcmVhdGVIb29rc0Ioe1xuICBhOiBcIiYuYVwiLFxuICBiOiBcIiYuYlwiLFxuICBjOiBcIiYuY1wiLFxuICBkOiBcIiYuZFwiLFxuICBlOiBcIiYuZVwiLFxuICBmOiBcIiYuZlwiLFxuICBnOiBcIiYuZ1wiLFxuICBoOiBcIiYuaFwiLFxuICBpOiBcIiYuaVwiLFxuICBqOiBcIiYualwiLFxuICBrOiBcIiYua1wiLFxuICBsOiBcIiYubFwiLFxuICBtOiBcIiYubVwiLFxuICBuOiBcIiYublwiLFxuICBvOiBcIiYub1wiLFxuICBwOiBcIiYucFwiLFxuICBxOiBcIiYucVwiLFxuICByOiBcIiYuclwiLFxuICBzOiBcIiYuc1wiLFxuICB0OiBcIiYudFwiLFxuICB1OiBcIiYudVwiLFxuICB2OiBcIiYudlwiLFxuICB3OiBcIiYud1wiLFxuICB4OiBcIiYueFwiLFxuICB5OiBcIiYueVwiLFxuICB6OiBcIiYuelwiLFxufSlbMV07IiwidGVzdHMiOlt7Im5hbWUiOiJBIC0gb3JpZ2luYWwiLCJjb2RlIjoiY3NzQSh7XG4gIGJhY2tncm91bmRDb2xvcjogXCJyZWRcIixcbiAgYm9yZGVyOiBcIjFweCBzb2xpZCBibGFja1wiLFxuICBhOiB7IGJhY2tncm91bmRDb2xvcjogXCJibHVlXCIgfVxufSk7XG5cbmNzc0EoXG4gIHsgY29sb3I6IFwiYmx1ZVwiLCBmb250U2l6ZTogXCIxNnB4XCIgfSxcbiAgeyBwYWRkaW5nOiBcIjEwcHhcIiwgYjogeyBjb2xvcjogXCJyZWRcIiB9IH1cbik7XG5cbmNzc0EoXG4gIHsgbWFyZ2luVG9wOiBcIjEwcHhcIiwgYTogeyBtYXJnaW5Ub3A6IFwiMjBweFwiIH0gfSxcbiAgeyBiOiB7IGNvbG9yOiBcImdyZWVuXCIsIGE6IHsgY29sb3I6IFwieWVsbG93XCIgfSB9IH1cbik7XG5cbmNzc0EoXG4gIHtcbiAgICBhOiB7XG4gICAgICBmb250V2VpZ2h0OiBcImJvbGRcIixcbiAgICAgIGI6IHsgZm9udFdlaWdodDogXCJub3JtYWxcIiB9XG4gICAgfVxuICB9LFxuICB7XG4gICAgYzoge1xuICAgICAgdGV4dERlY29yYXRpb246IFwidW5kZXJsaW5lXCIsXG4gICAgICBkOiB7IHRleHREZWNvcmF0aW9uOiBcIm5vbmVcIiB9XG4gICAgfVxuICB9LFxuICB7XG4gICAgZToge1xuICAgICAgZm9udFN0eWxlOiBcIml0YWxpY1wiLFxuICAgICAgZjogeyBmb250U3R5bGU6IFwibm9ybWFsXCIgfVxuICAgIH1cbiAgfVxuKTtcblxuY3NzQShcbiAge1xuICAgIGJhY2tncm91bmRDb2xvcjogXCJ5ZWxsb3dcIixcbiAgICBiOiB7IGJhY2tncm91bmRDb2xvcjogXCJvcmFuZ2VcIiB9XG4gIH0sXG4gIHtcbiAgICBhOiB7XG4gICAgICBtYXJnaW5MZWZ0OiBcIjIwcHhcIixcbiAgICAgIGI6IHsgbWFyZ2luTGVmdDogXCIzMHB4XCIgfVxuICAgIH0sXG4gICAgYzoge1xuICAgICAgZm9udFNpemU6IFwiMTRweFwiLFxuICAgICAgZDogeyBmb250U2l6ZTogXCIxNnB4XCIgfVxuICAgIH1cbiAgfVxuKTsiLCJydW5zIjpbMzUwMCwzNjY2LDczMzMsNTAwMCw3MzMzLDY1MDAsNTE2Niw1NTAwLDYwMDAsNjAwMCw1NjY2LDUxNjYsNzgzMyw1MTY2LDU2NjYsNTMzMywzMTY2LDY1MDAsNTAwMCw2MzMzLDYzMzMsNjE2Niw1MDAwLDQxNjYsNjY2Niw1ODMzLDY1MDAsNTY2Niw1NTAwLDcwMDAsNjAwMCw1MzMzLDU4MzMsNjUwMCw2ODMzLDcwMDAsNjUwMCw1MzMzLDY2NjYsNzAwMCw1MzMzLDY1MDAsNTUwMCw2MzMzLDUwMDAsNDY2Niw0ODMzLDQ4MzMsNDgzMyw1MzMzLDY4MzMsNzUwMCw1MTY2LDYzMzMsNzE2Niw1MzMzLDcxNjYsNjUwMCw1MTY2LDcwMDAsNDgzMyw2MTY2LDY1MDAsNjUwMCw1NjY2LDU4MzMsNTMzMyw2NjY2LDY1MDAsNzUwMCw1MzMzLDYwMDAsNjY2Niw1NjY2LDU4MzMsNTAwMCw0ODMzLDYwMDAsNjgzMyw2NjY2LDYwMDAsNTgzMywzODMzLDUxNjYsNTAwMCw1MTY2LDcxNjYsNjMzMyw1MzMzLDUwMDAsNjMzMyw1ODMzLDY4MzMsNTAwMCw1MDAwLDY2NjYsNDUwMCw2NjY2LDU1MDAsNTE2Nl0sIm9wcyI6NTgzNn0seyJuYW1lIjoiQiAtIGltcGVyYXRpdmUgc3R5bGUiLCJjb2RlIjoiY3NzQih7XG4gIGJhY2tncm91bmRDb2xvcjogXCJyZWRcIixcbiAgYm9yZGVyOiBcIjFweCBzb2xpZCBibGFja1wiLFxuICBhOiB7IGJhY2tncm91bmRDb2xvcjogXCJibHVlXCIgfVxufSk7XG5cbmNzc0IoXG4gIHsgY29sb3I6IFwiYmx1ZVwiLCBmb250U2l6ZTogXCIxNnB4XCIgfSxcbiAgeyBwYWRkaW5nOiBcIjEwcHhcIiwgYjogeyBjb2xvcjogXCJyZWRcIiB9IH1cbik7XG5cbmNzc0IoXG4gIHsgbWFyZ2luVG9wOiBcIjEwcHhcIiwgYTogeyBtYXJnaW5Ub3A6IFwiMjBweFwiIH0gfSxcbiAgeyBiOiB7IGNvbG9yOiBcImdyZWVuXCIsIGE6IHsgY29sb3I6IFwieWVsbG93XCIgfSB9IH1cbik7XG5cbmNzc0IoXG4gIHtcbiAgICBhOiB7XG4gICAgICBmb250V2VpZ2h0OiBcImJvbGRcIixcbiAgICAgIGI6IHsgZm9udFdlaWdodDogXCJub3JtYWxcIiB9XG4gICAgfVxuICB9LFxuICB7XG4gICAgYzoge1xuICAgICAgdGV4dERlY29yYXRpb246IFwidW5kZXJsaW5lXCIsXG4gICAgICBkOiB7IHRleHREZWNvcmF0aW9uOiBcIm5vbmVcIiB9XG4gICAgfVxuICB9LFxuICB7XG4gICAgZToge1xuICAgICAgZm9udFN0eWxlOiBcIml0YWxpY1wiLFxuICAgICAgZjogeyBmb250U3R5bGU6IFwibm9ybWFsXCIgfVxuICAgIH1cbiAgfVxuKTtcblxuY3NzQihcbiAge1xuICAgIGJhY2tncm91bmRDb2xvcjogXCJ5ZWxsb3dcIixcbiAgICBiOiB7IGJhY2tncm91bmRDb2xvcjogXCJvcmFuZ2VcIiB9XG4gIH0sXG4gIHtcbiAgICBhOiB7XG4gICAgICBtYXJnaW5MZWZ0OiBcIjIwcHhcIixcbiAgICAgIGI6IHsgbWFyZ2luTGVmdDogXCIzMHB4XCIgfVxuICAgIH0sXG4gICAgYzoge1xuICAgICAgZm9udFNpemU6IFwiMTRweFwiLFxuICAgICAgZDogeyBmb250U2l6ZTogXCIxNnB4XCIgfVxuICAgIH1cbiAgfVxuKTsiLCJydW5zIjpbMzY2Niw1ODMzLDc4MzMsNTE2Niw3ODMzLDcwMDAsNTUwMCw3MTY2LDc4MzMsNjMzMyw2MzMzLDY1MDAsODY2Niw2NjY2LDY1MDAsODMzMywzMzMzLDc4MzMsNjUwMCw2MTY2LDcwMDAsNjgzMyw2MzMzLDUwMDAsNzgzMyw2ODMzLDU2NjYsNjMzMyw3ODMzLDYzMzMsNTMzMyw3MzMzLDY1MDAsNzAwMCw3MzMzLDc4MzMsNjY2Niw2ODMzLDY1MDAsNjE2Niw1NjY2LDczMzMsNzAwMCw3MDAwLDYzMzMsNjUwMCw2ODMzLDYzMzMsNzE2Niw1MzMzLDYzMzMsNjgzMyw2MzMzLDYxNjYsODgzMyw1MzMzLDY1MDAsNjUwMCw3MDAwLDY1MDAsNjY2Niw3MDAwLDYzMzMsNzMzMyw1MzMzLDY1MDAsNjY2Niw1MTY2LDcwMDAsNzUwMCw2MzMzLDc4MzMsNjY2Niw2ODMzLDgwMDAsNTgzMyw1ODMzLDczMzMsNzUwMCw3MzMzLDc4MzMsNzMzMyw0MTY2LDczMzMsNzAwMCw3MDAwLDY1MDAsNzE2Niw4MDAwLDUzMzMsODMzMyw1MzMzLDY4MzMsNTAwMCw2MzMzLDcwMDAsNjAwMCw2MzMzLDczMzMsNTUwMF0sIm9wcyI6NjYyOX1dLCJ1cGRhdGVkIjoiMjAyNC0wMS0wNVQyMDo1MjoyMC4xMDVaIn0%3D